### PR TITLE
Reduce record usage in published projects

### LINKED ISSUE #45859: Reduce record usage in published aspnetcore projects
records have a slim syntax and provide many features, but they also generate a lot of code. And much of that code can't be trimmed. A record can bloat app publish size.

See https://github.com/dotnet/aspnetcore/pull/45604#discussion_r1061055000 for an example.

- [x] Audit published aspnetcore projects for record usage. records in test projects is fine.
- [ ] See if there is an analyzer that we can enable to warn about record use, and enable.

### PR DESCRIPTION
**Pull request by david-acker**
### Reduce record usage in published projects

Converts instances of `record` to `class` and `record struct` to `struct`.

Fixes #45859

### CODE REVIEW

**Comment by david-acker** [comment]
FYI I'll be taking a quick look through the projects again just to make sure I didn't miss anything.

**Review by captainsafia — APPROVED** [review]

**Comment by adityamandaleeka** [comment]
Thanks @david-acker! Looks like you have some merge conflicts.

**Review by captainsafia — APPROVED** [review]
LGTM!

Would be good to get a look from @dotnet/aspnet-blazor-eng  to cover the changes under `src/Components` and @amcasey for review on the `src/Server` changes.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs [hunk 1]
```diff
@@ -598,4 +598,15 @@ private static bool IsHex(char ch)
     }
 }
 
-internal sealed record AltSvcHeader(string Value, byte[] RawBytes);
+internal sealed class AltSvcHeader
```
Comment (amcasey) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs:601]: I couldn't find any places where we were using logical equality or hashcode.

File: src/Servers/Kestrel/shared/KnownHeaders.cs [hunk 2]
```diff
@@ -40,23 +40,37 @@ public class KnownHeaders
...
-    // These headers are excluded from generated IHeadersDictionary implementaiton.
+    // These headers are excluded from generated IHeadersDictionary implementation.
     public static readonly string[] NonPublicHeaderNames = new[]
     {
         HeaderNames.DNT,
         InternalHeaderNames.AltUsed
     };
 
-    public record InternalHeader(string Identifier, string Name, bool IsPseudoHeader = false);
+    public class InternalHeader
```
Comment (amcasey) [inline src/Servers/Kestrel/shared/KnownHeaders.cs:50]: Sealed?

Comment (amcasey) [inline src/Servers/Kestrel/shared/KnownHeaders.cs:50]: I don't think this "shared" means "with runtime", but @BrennanConroy might be able to confirm.

**Review by amcasey — APPROVED** [review]
Makes sense to me (the Server folder).

File: src/Servers/Kestrel/shared/KnownHeaders.cs (hunk 2 above)
Comment (david-acker) [reply src/Servers/Kestrel/shared/KnownHeaders.cs]: Sealed!

Comment (BrennanConroy) [reply src/Servers/Kestrel/shared/KnownHeaders.cs]: I don't understand the question

Comment (amcasey) [reply src/Servers/Kestrel/shared/KnownHeaders.cs]: We have some code that is shared with the runtime repo.  This isn't part of it, correct?

Comment (BrennanConroy) [reply src/Servers/Kestrel/shared/KnownHeaders.cs]: Right. That code is in https://github.com/dotnet/aspnetcore/tree/main/src/Shared/runtime

**Comment by adityamandaleeka** [comment]
@dotnet/aspnet-blazor-eng Can someone please take a quick look at the `src/Components` changes so we can get this in for .NET 8?

**Review by javiercn — APPROVED** [review]

**Comment by david-acker** [comment]
Could someone with commit rights use `/azp run`?

**Comment by david-acker** [comment]
Other changes:
- Converted `SequenceAndKey`(which was recently added to `EndpointHtmlRenderer.Streaming.cs`) from a record struct to a struct.
- Re-converted `FormDataConverterReadParameters` (from `ComplextTypeExpressionConverterFactoryOfT.cs`) to a struct. It looks like this was moved to a different file recently and didn't retain the original changes.

**Comment by adityamandaleeka** [comment]
Thanks @david-acker!

**Pull request merged**